### PR TITLE
fix: resolve view dependency for create_mv and create_view

### LINKED ISSUE #6160: Resolve logical view's dependants
Original posted at https://github.com/risingwavelabs/risingwave/pull/6023

View's dependants not resolved. e.g., when view `V` depends on mv (or table) `MV`, we can't drop `MV` before dropping `V`. But if mv (or view) `MV` depends on `V`, we can drop `V`. 
  
The difficulty is that currently after binding, view is completely replaced by the sql and the information of the view is lost. To support it, we may collect the dependent view ids in binder. Note that currently for mv, dependent relation ids are resolved in meta from stream fragment plan. For view, they are resolved in frontend from batch plan.

### PR DESCRIPTION
**Pull request by yezizp2012**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Follow up for #7745 , resolving dependency information of views for `create_mv` and `create_view`, and source dependency for `create_view`, to correctly support drop check and recursive rename for view.

## Checklist For Contributors

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [x] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [ ] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

- [x] My PR **DOES NOT** contain user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>
Resolve #6160

### CODE REVIEW

File: src/frontend/src/handler/create_mv.rs [hunk 1]
```diff
@@ -86,15 +86,15 @@ pub fn gen_create_mv_plan(
...
         let bound = binder.bind_query(query)?;
-        // FIXME: We should record the views into mv's dependent relations to to refine the drop
-        // check and recursive rename of the views.
-        let _views = binder.shared_views().keys().cloned().collect_vec();
-        bound
+        (binder.shared_views(), bound)
     };
 
+    let check_items = resolve_query_privileges(&bound);
+    session.check_privileges(&check_items)?;
```
Comment (xxchan) [inline src/frontend/src/handler/create_mv.rs:96]: What's this change about? 👀

File: src/frontend/src/handler/query.rs [hunk 2]
```diff
@@ -90,16 +91,24 @@ fn must_run_in_local_mode(bound: &BoundStatement) -> bool {
     must_local
 }
 
+pub struct BatchQueryPlanResult {
+    pub(crate) plan: PlanRef,
+    pub(crate) query_mode: QueryMode,
+    pub(crate) schema: Schema,
+    pub(crate) dependent_views: Vec<ViewId>,
```
Comment (xxchan) [inline src/frontend/src/handler/query.rs:98]: It looks a little bit confusing why we treat `dependent_views` differently here. 🥹

Is it possible to resolve all relations here? (Ok to add a TODO) If not or hard, maybe add some comments to explain the reason.

File: src/frontend/src/handler/create_mv.rs [hunk 3]
```diff
@@ -86,15 +86,15 @@ pub fn gen_create_mv_plan(
...
 
-    let bound = {
+    let (dependent_views, bound) = {
         let mut binder = Binder::new_for_stream(session);
         let bound = binder.bind_query(query)?;
-        // FIXME: We should record the views into mv's dependent relations to to refine the drop
-        // check and recursive rename of the views.
-        let _views = binder.shared_views().keys().cloned().collect_vec();
-        bound
+        (binder.shared_views(), bound)
```
Comment (xxchan) [inline src/frontend/src/handler/create_mv.rs:92]: I feel treating views separately here might be confusing and error-prone.. Similar concern with the comment below on `BatchQueryPlanResult`

File: src/frontend/src/handler/create_mv.rs (hunk 1 above)
Comment (yezizp2012) [reply src/frontend/src/handler/create_mv.rs:96]: The logic for checking privileges was missing here previously.

File: src/frontend/src/handler/query.rs (hunk 2 above)
Comment (yezizp2012) [reply src/frontend/src/handler/query.rs:98]: Yes exactly. I'm thinking of putting all the resolving of dependency information together on the frontend. Part of it is currently being done in fragmenter. I will add a TODO here and fix it in another PR.

**Review by xx01cyx — COMMENTED** [review]
Generally LGTM. The same feeling as @xxchan: we should unify resolving dependent relations instead of separating views out.

File: src/frontend/src/handler/query.rs (hunk 2 above)
Comment (yezizp2012) [reply src/frontend/src/handler/query.rs]: I unified resolving dependent relations in the frontend while binding `table`/`view`/`source`. PTAL. Cc @xx01cyx

File: src/connector/src/sink/catalog/desc.rs [hunk 4]
```diff
@@ -61,7 +61,7 @@ impl SinkDesc {
         schema_id: SchemaId,
         database_id: DatabaseId,
         owner: UserId,
-        dependent_relations: Vec<TableId>,
+        dependent_relations: Vec<u32>,
```
Comment (BugenZhao) [inline src/connector/src/sink/catalog/desc.rs:64]: What's the semantics of `u32` here? I guess we're de facto using `TableId` as the general type for identifying all relations. 🥵

Comment (yezizp2012) [reply src/connector/src/sink/catalog/desc.rs:64]: The semantic is relation. 🥵 yes we need to introduce a general type for identifying all relations and generate ids for all of them from the same category in meta also(currently they all use `Table`).

Comment (yezizp2012) [reply src/connector/src/sink/catalog/desc.rs]: Reverted. Will do some refactor on the ids.

File: proto/stream_plan.proto [hunk 5]
```diff
@@ -689,7 +689,7 @@ message StreamFragmentGraph {
   // edges between fragments.
   repeated StreamFragmentEdge edges = 2;
 
-  repeated uint32 dependent_relation_ids = 3;
+  repeated uint32 dependent_table_ids = 3;
```
Comment (xx01cyx) [inline proto/stream_plan.proto:692]: Does the renaming indicate that the IDs are for tables/mviews or that the structs are `TableId`s?

Comment (yezizp2012) [reply proto/stream_plan.proto:692]: Yes, the ids here are all streaming job ids. Previously the field is also used for dependency resolving and source ids were included as a workaround.

**Review by xxchan — COMMENTED** [review]
The problem I mentioned here occurs 😇  https://github.com/risingwavelabs/risingwave/issues/6160#issuecomment-1310528061 

```
set rw_streaming_enable_delta_join = true;
create table a (a1 int, a2 int);
create index i_a1 on a(a1);
create table b (b1 int, b2 int);
create index i_b1 on b(b1);
create materialized view mv as select * from a join b on a.a1 = b.b1 ;
drop index i_a1;

-- server closed the connection unexpectedly
```

**Comment by yezizp2012** [comment]
> The problem I mentioned here occurs 😇 [#6160 (comment)](https://github.com/risingwavelabs/risingwave/issues/6160#issuecomment-1310528061)
> 
> i.e., if we resolve dependencies syntactically in binder, implicit dependencies on indices might got lost.
> 
> ```
> set rw_streaming_enable_delta_join = true;
> create table a (a1 int, a2 int);
> create index i_a1 on a(a1);
> create table b (b1 int, b2 int);
> create index i_b1 on b(b1);
> create materialized view mv as select * from a join b on a.a1 = b.b1 ;
> drop index i_a1;
> 
> -- server closed the connection unexpectedly
> ```

Then we have to find all referenced relations by a plan visitor, which is the same as what was done in `create_view`. But the referenced views can only be resolved in binder.  😇 I will find a less-confusing way to unify them.

> BTW, this might also affect the behavior of this:
> 
> ```
> create materialized view mv as 
>          with ctx as (select x from t1) select x from t2; drop table t1;
> ```
> 
> Maybe need some more tests.

Table `t1` will be recorded in binder, so it wouldn't be affected. But we change to plan visitor way, it will be left out. So we might need them both.🤔

File: e2e_test/ddl/dependency_check.slt [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+drop source src;
+
+statement ok
+drop materialized view mv2;
+
+statement ok
+drop table a;
+
+statement ok
+drop table b;
```
Comment (xxchan) [inline e2e_test/ddl/dependency_check.slt:79]: Wait, i_a1 and i_b1 are not dropped... 🫠

Comment (yezizp2012) [reply e2e_test/ddl/dependency_check.slt:79]: The index will be dropped together with the table. This is the same behavior as PostgreSQL.

**Review by xxchan — APPROVED** [review]
LGTM 😭😭😭

File: src/frontend/src/optimizer/plan_visitor/relation_collector_visitor.rs [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+impl RelationCollectorVisitor {
+    fn new_with(relations: HashSet<TableId>) -> Self {
+        Self { relations }
+    }
+
+    /// `collect_with` will collect all the relations in the plan with some default ones, which are
+    /// collected during the binding phase. Note that during visit the collected relations might be
+    /// duplicated with the default ones.
+    pub fn collect_with(relations: HashSet<TableId>, plan: PlanRef) -> HashSet<TableId> {
```
[the remark this answers is not in the document]
Comment (yezizp2012) [reply src/frontend/src/optimizer/plan_visitor/relation_collector_visitor.rs:37]: 😭 comments added.

**Pull request merged**